File: networkmodule/src/main/java/com/test/networkmodule/backend/ShoppingReceiver.java
package com.test.networkmodule.backend;

import android.text.TextUtils;

import com.test.networkmodule.exceptions.ApiException;
import com.test.networkmodule.response.MasterListResponse;
import com.test.networkmodule.response.ShopListingResponse;
import com.test.networkmodule.retrofitapi.ShoppingRetrofitAPI;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Single;

@Singleton
public class ShoppingReceiver {

    private final ShoppingRetrofitAPI shoppingRetrofitAPI;

    private final static String SHOPPING_MASTER_UL = "https://s3-ap-northeast-1.amazonaws.com/m-et/Android/json/master.json";

    @Inject
    public ShoppingReceiver(ShoppingRetrofitAPI shoppingRetrofitAPI) {
        this.shoppingRetrofitAPI = shoppingRetrofitAPI;
    }

    public Single<List<ShopListingResponse>> getShoppingList(String name) {
        return shoppingRetrofitAPI.getMasterListResponse(SHOPPING_MASTER_UL)
                .map(response -> {
                    if (response.isSuccessful()) {
                        return response.body();
                    } else {
                        throw new ApiException("Master API Failure");
                    }
                })
                .map(masterListResponse -> getUrlForName(masterListResponse, name))
                .map(url -> {
                    if (TextUtils.isEmpty(url)) {
                        throw new ApiException("Content not available");
                    } else {
                        return url;
                    }
                })
                .flatMap(shoppingRetrofitAPI::getShoppingList)
                .map(shopListingResponseResponse -> {
                    if (shopListingResponseResponse.isSuccessful()) {
                        return shopListingResponseResponse.body();
                    } else {
                        throw new ApiException(name + " API Failure");
                    }
                });
    }


    private String getUrlForName(List<MasterListResponse> masterListResponseList, String name) {
        for (MasterListResponse masterResponse : masterListResponseList) {
            if (masterResponse.name().equalsIgnoreCase(name)) {
                return masterResponse.data();
            }
        }
        return null;
    }
}

File: networkmodule/src/main/java/com/test/networkmodule/backend/ShoppingReceiver.kt
package com.test.networkmodule.backend

import android.text.TextUtils
import com.test.networkmodule.exceptions.ApiException
import com.test.networkmodule.response.MasterListResponse
import com.test.networkmodule.response.ShopListingResponse
import com.test.networkmodule.retrofitapi.ShoppingRetrofitAPI
import io.reactivex.Single
import retrofit2.Response
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ShoppingReceiver @Inject
constructor(private val shoppingRetrofitAPI: ShoppingRetrofitAPI) {

    fun getShoppingList(name: String): Single<List<ShopListingResponse>> {

        return shoppingRetrofitAPI.getMasterListResponse(SHOPPING_MASTER_UL)
            .map { t -> getMasterResponse(t) }
            .map { t -> getUrlForName(t, name) }
            .map { t -> checkValidUrl(t) }
            .flatMap { t -> shoppingRetrofitAPI.getShoppingList(t) }
            .map { t -> getShoppingListResponse(t) }

    }

    private fun getShoppingListResponse(response: Response<List<ShopListingResponse>>): List<ShopListingResponse> {
        return when {
            response.isSuccessful -> response.body()
            else -> throw ApiException("Content not available")
        }
    }

    private fun getMasterResponse(response: Response<List<MasterListResponse>>): List<MasterListResponse> {
        return when {
            response.isSuccessful -> response.body()
            else -> throw ApiException("Master API Failure")
        }
    }

    private fun checkValidUrl(url: String): String {
        return when {
            TextUtils.isEmpty(url) -> throw ApiException("Content not available")
            else -> url
        }
    }

    private fun getUrlForName(
        masterListResponseList: List<MasterListResponse>,
        name: String
    ): String? {
        for ((name1, data) in masterListResponseList) {
            if (name1.equals(name, ignoreCase = true)) {
                return data
            }
        }
        return null
    }

    companion object {
        private const val SHOPPING_MASTER_UL =
            "https://s3-ap-northeast-1.amazonaws.com/m-et/Android/json/master.json"
    }
}
